c++/src/kj/compat/tls.c++

```diff
@@ -405,31 +416,108 @@ private:
 // Implementations of ConnectionReceiver, NetworkAddress, and Network as wrappers adding TLS.
 
-class TlsConnectionReceiver final: public kj::ConnectionReceiver {
+template <typename T>
+class InputQueue {
+  // InputQueue is an async FIFO queue that has many sources but only one sink.
+  // TODO(pr) Should I find a header home for this somewhere?
+
 public:
-  TlsConnectionReceiver(TlsContext& tls, kj::Own<kj::ConnectionReceiver> inner)
-      : tls(tls), inner(kj::mv(inner)) {}
+  void pushRaw(T&& v) {
+    // Push an existing value onto the queue.
 
-  Promise<Own<AsyncIoStream>> accept() override {
-    return inner->accept().then([this](kj::Own<AsyncIoStream> stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return accept();
-      });
+    if (auto localFulfiller = kj::mv(fulfiller)) {
+      // Fast path: We have a consumer waiting.
+      localFulfiller->fulfill(kj::fwd<T>(v));
+      return;
+    }
+
+    // Slow path: We don't have a consumer, add it to the queue.
+    auto node = kj::heap<Node>(kj::fwd<T>(v));
+    if (head) {
+      Node* currentNode = head.get();
+      while (currentNode->next) {
+        // Run through the nodes until we don't have a next.
+        currentNode = currentNode->next.get();
+      }
+
+      currentNode->next = kj::mv(node);
+    } else {
+      head = kj::mv(node);
+    }
+  }
+
+  Promise<void> push(Promise<T>&& p) {
+    // Eventually push a value onto the queue.
+
+    return kj::fwd<Promise<T>>(p).then([this](T&& v) {
+      pushRaw(kj::fwd<T>(v));
     });
   }
 
-  Promise<kj::AuthenticatedStream> acceptAuthenticated() override {
-    return inner->acceptAuthenticated().then([this](kj::AuthenticatedStream stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return acceptAuthenticated();
-      });
+  Promise<T> pop() {
+    // Eventually pop a value from the queue.
+    // Note that if your sink lags your sources, the promise will always be ready.
+
+    KJ_REQUIRE(!fulfiller, "InputQueue can only have one waiter");
+
+    if (head) {
+      // Fast path: There are already values in queue for us to take.
+      auto value = kj::mv(head->value);
+      ;
+      head = kj::mv(head->next);
+      return kj::mv(value);
+    } else {
+      // Slow path: We need to wait for a value to arrive.
+      auto paf = newPromiseAndFulfiller<T>();
+      fulfiller = kj::mv(paf.fulfiller);
+
+      return kj::mv(paf.promise);
+    }
+  }
+
+private:
+  struct Node {
+    // A simple intrusive linked list node which extends the lifetime of the next node.
+
+    Node(Promise<T>&& value): next{}, value{ kj::fwd<Promise<T>>(value) } {
```

Comment:
It looks like when you actually construct `Node`s it's always with a value of type `T`, not a `Promise`, so wrapping in a `Promise` is redundant and adds overhead.

Reply:
Yeah, you're right. Thanks for the catch.